backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,287 @@
+"""Contains derivatives for BatchNorm."""
+from typing import Dict, List, Tuple, Union
+
+from torch import Size, Tensor, einsum
+from torch.nn import BatchNorm1d, BatchNorm2d, BatchNorm3d
+
+from backpack.core.derivatives.basederivatives import BaseParameterDerivatives
+
+
+class BatchNormNdDerivatives(BaseParameterDerivatives):
+    """Derivatives for BatchNorm1d, 2d and 3d.
+
+    If training=False: saved statistics are used.
+    If training=True: statistics of current batch are used.
+
+    Index convention:
+    n: batch axis
+    c: category axis
+    {empty}/l/hw/dhw: dimension axis for 0/1/2/3-dimensions
+
+    Links to PyTorch docs:
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm3d.html
+
+    As a starting point for derivative computation, see these references:
+    https://kevinzakka.github.io/2016/09/14/batch_normalization/
+    https://chrisyeh96.github.io/2017/08/28/deriving-batchnorm-backprop.html
+    """
+
+    def _check_parameters(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> None:
+        if module.affine is False:
+            raise NotImplementedError("Only implemented for affine=True")
+        if module.track_running_stats is False:
+            raise NotImplementedError("Only implemented for track_running_stats=True")
+
+    def hessian_is_zero(self) -> bool:
+        """Whether hessian is zero.
+
+        Returns:
+            whether hessian is zero
+        """
+        return False
```

Comment:
I believe this depends on the mode: With `training=True`, the output is nonlinear in the input (Hessian non-zero), but when `training=False`, then the output is linear in the input, hence the Hessian should be zero.

I just merged testing functionality for `hessian_is_zero` into `development` (#183). You could either merge this into your branch and verify what I wrote by adding a test. Alternatively, you can assume that the old code assumed `training=False` and throw a `NotImplementedError` if `training=True`.

Reply:
Since the interface is `hessian_is_zero(self)`, there is no way of determining the training mode. To determine the training mode I need `module.training`.

Reply:
Solutions:
- leave as it is
- change entire interface
- always raise error

Reply:
Damn. I also realized this [here](https://github.com/fKunstner/backpack-discuss/issues/110).

The solution will be to modify the interface. I believe leaving it as is is dangerous because we will forget about it. Always raising an error is also not possible because this will make some tests fail.

I think we are kind of about to get stuck with batch norm. Maybe we should take one step back, and improve the interface before coming back. Added to the agenda for Tuesday.

Reply:
changed interface